File: server/models/bullet.py
from __future__ import annotations
from typing import TYPE_CHECKING
from math import sin, cos
from attrs import astuple
from dto import BulletDataDto
from utils.uid import CID
from .vector import Vector
from .entity import Entity

if TYPE_CHECKING:
    from rooms.game_states import PlayingGameState

BULLET_SPEED = 1024
RADIUS_GROWTH_TEMPO = 64
MAX_BULLET_RADIUS = 16
GRAVITY_PULL = Vector(0, -200, 0)


class Bullet(Entity):
    def __init__(
        self,
        *,
        world: PlayingGameState,
        owner: CID,
        direction: float,
        pos: Vector
    ):
        super().__init__(
            world=world,
            pos=pos,
            vel=Vector(sin(direction), 0, cos(direction)) * BULLET_SPEED,
            radius=0
        )
        self.owner = owner

    def calculate_forces(self) -> Vector:
        return GRAVITY_PULL * self._mass

    def update(self, dtime: float):
        super().update(dtime)
        self.radius = min(self.radius + dtime * RADIUS_GROWTH_TEMPO, MAX_BULLET_RADIUS)

    def to_dto(self) -> BulletDataDto:
        return BulletDataDto(
            self.eid, self.owner, astuple(self.pos), self.radius
        )

File: Backend/src/Domain/Game/Bullet.cs
using Backend.Domain.Identifiers;

namespace Backend.Domain.Game;

internal sealed class Bullet : Entity
{
    private const double BulletSpeed = 10; // m/s
    private const double RadiusGrowthTempo = 0.5; // m/s
    private const double MaxBulletRadius = 0.15; // m

    public Bullet(
        IWorld world,
        Cid ownerCid,
        double direction,
        Vector position) : base(
        world,
        position: position,
        velocity: Vector.UnitWithPitch(direction) * BulletSpeed,
        radius: 0) =>
        OwnerCid = ownerCid;

    public Cid OwnerCid { get; }

    protected override Vector CalculateForces() => IWorld.Gravity * Mass;

    public override void Update(TimeSpan deltaTime)
    {
        base.Update(deltaTime);
        Radius = Math.Min(Radius + RadiusGrowthTempo * deltaTime.TotalSeconds, MaxBulletRadius);
    }
}
